src/collective/elasticsearch/tests/test_search.py

```diff
@@ -69,180 +59,283 @@ def test_date_index_query(self):
 
-    def test_brains_indexing(self):
+
+class TestBrainsIndexing(BaseFunctionalTest):
+    def setUp(self):
+        super().setUp()
         for idx in range(120):
-            createObject(self.portal, "Document", f"{idx:04d}page", title=f"Page {idx}")
+            api.content.create(
+                self.portal, "Document", f"{idx:04d}page", title=f"Page {idx}"
+            )
+        self.commit(wait=1)
+        self.el_results = self.search(
+            {
+                "portal_type": "Document",
+                "sort_on": "getId",
+                "sort_order": "asc",
+            }
+        )
+
+    def test_all_indexed(self):
+        self.assertEqual(len(self.el_results), 120)
+
+    @parameterized.expand(
+        [
+            (0, "0000page"),
+            (-1, "0119page"),
+            (-50, "0070page"),
+            (-55, "0065page"),
+            (-100, "0020page"),
+        ]
+    )
+    def test_ordering(self, result_idx, expected):
+        self.assertEqual(self.el_results[result_idx].getId, expected)
+
+
+class TestCatalogRecordDeleted(BaseFunctionalTest):
+    def setUp(self):
+        super().setUp()
+        catalog = self.catalog._catalog
+        self.event = api.content.create(
+            self.portal, "Event", "event-test", title="Gone Event"
+        )
+        self.commit(wait=1)
+        path = "/".join(self.event.getPhysicalPath())
+        catalog.uncatalogObject(path)
         self.commit()
-        el_results = self.catalog(
-            portal_type="Document", sort_on="getId", sort_order="asc"
+
+    def test_search_results(self):
+        el_results = self.search({"portal_type": "Event", "Title": "Gone Event"})
+        self.assertEqual(len(el_results), 1)
+        brain = el_results[0]
+        self.assertTrue(ICatalogBrain.providedBy(brain))
+        self.assertEqual(brain.getRID(), -1)
+        # Test data from elastic will populate the brain
+        self.assertEqual(brain.portal_type, "Event")
+        self.assertEqual(brain.Title, "Gone Event")
+        # Test
+        self.assertEqual(brain.getPath(), "/plone/event-test")
+        self.assertEqual(brain.getURL(), "http://nohost/plone/event-test")
+        self.assertEqual(brain.getObject(), self.event)
+
+
+class TestDeleteObjectNotReflectedOnES(BaseFunctionalTest):
+    def setUp(self):
+        super().setUp()
+        catalog = self.catalog._catalog
+        self.event = api.content.create(
+            self.portal, "Event", "event-test", title="Gone Event"
         )
-        self.assertEqual(len(el_results), 120)
-        self.assertEqual(el_results[0].getId, "0000page")
-        self.assertEqual(el_results[-1].getId, "0119page")
-        self.assertEqual(el_results[-50].getId, "0070page")
-        self.assertEqual(el_results[-55].getId, "0065page")
-        self.assertEqual(el_results[-100].getId, "0020page")
+        self.commit(wait=1)
+        path = "/".join(self.event.getPhysicalPath())
+        catalog.uncatalogObject(path)
```

Comment:
> Just double checking, uncatalog does remove the record from ElasticSearch

I see that in this line the `uncatalogObject` is called with the path of the Event. But in the test_search_results below, the search finds the Event:

https://github.com/collective/collective.elasticsearch/blob/ca4f83b73ecbe4635018accdc0bcfac5db33cc43/src/collective/elasticsearch/tests/test_search.py#L286-L288

What am I missing?

Reply:
@wesleybl This is the expected behavior. There is an object that is not on portal_catalog but still is in ES, so the result should not break during access to the result.
The uncatalogObject is used to simulate this scenario

Reply:
Before that, accessing el_results[0] would return None and break plone.api

Reply:
> This is the expected behavior.

I don't think this is the expected behavior. When calling `uncatalogObject`, I expect the record to be removed from Elastic. Even in the master and previus versions it is like this. See:

https://github.com/collective/collective.elasticsearch/blob/0375accc03093c81f16355f773b53b83f317cb57/src/collective/elasticsearch/patches/__init__.py#L13-L15

But you changed this behavior in this PR.

Probably, the situation that occurred of the record being in Elastic and not being in the catalog was due to some error. But to simulate this problem in the test, you are making every call to `uncatalogObject` "generate this error".

I think the error simulation should only be done in the test, not in every `uncatalogObject` call, which can occur outside of the test. In a comment above, I gave a suggestion on how to simulate the error only in the test, and keeping the `uncatalogObject` removing record from Elastic.

Reply:
@wesleybl, now I understand where is our "misunderstanding.":

tl;dr: Removing an object from the catalog should remove the object from Elastic Search using `QueueProcessor.unindex`. The call in the test is to simulate a case where the removal in Elastic Search failed.

Uncataloging the object now happens in [QueueProcessor](https://github.com/collective/collective.elasticsearch/blob/use-queue-processor/src/collective/elasticsearch/queueprocessor.py#L116), and this is called by [CMFCore](https://github.com/zopefoundation/Products.CMFCore/blob/master/src/Products/CMFCore/indexing.py#L228).

Portal catalog also implements un-indexing using a [QueueProcessor](https://github.com/zopefoundation/Products.CMFCore/blob/master/src/Products/CMFCore/indexing.py#L54), and in there you can see the data being removed from the catalog.

What I added to that test is a call to [Catalog.unindexObject](https://github.com/zopefoundation/Products.ZCatalog/blob/master/src/Products/ZCatalog/Catalog.py#L380) that will remove the object from the catalog **without** triggering any additional event or call. 

The idea of the test is to simulate a scenario where:

- User deletes an object on the site
- Transaction happens
- QueueProcessor for the Portal Catalog is effective
- QueueProcessor for ElasticSearch fails to erase the record
- A new query happens, and the *bad* record is returned

Reply:
> What I added to that test is a call to [Catalog.unindexObject](https://github.com/zopefoundation/Products.ZCatalog/blob/master/src/Products/ZCatalog/Catalog.py#L380) that will remove the object from the catalog without triggering any additional event or call.

@ericof  about using the `unindexObject` method test us ok. I understood!

But what I still find strange is when I go to the Catalog tab of portal_catalog and order a record to be removed. It is removed from portal_catalog but still in Elastic. In this situation I expect the registration to be removed from Elastic. This action calls the `uncatalog_object` method, which you remove patch in this PR. To see:

https://github.com/zopefoundation/Products.ZCatalog/blob/6885451bc4f4313dba970c4880e610f922e0f13b/src/Products/ZCatalog/ZCatalog.py#L238

Reply:
@wesleybl After sleeping on it, I realized the call via ZMI would not trigger QueueProcessor (as expected), so I'm adding the patch back.

Reply:
@wesleybl Fixed, and added tests